elpy-shell.el

```diff
@@ -884,71 +915,118 @@ below point and send the group around this statement."
       (goto-char (point-max)))
     (setq mark-active nil)))
 
-(defun elpy-shell-send-region-or-buffer-and-step ()
-  "Send the active region or the buffer to the Python shell."
+(defun elpy-shell-send-codecell-and-step ()
+  "Send the current code cell to the Python shell and step.
+
+Signals an error if the point is not inside a code cell.
+
+Cell beginnings and cell boundaries can be customized via the
+variables `elpy-shell-cell-boundary-regexp' and
+`elpy-shell-codecell-beginning-regexp', which see."
   (interactive)
+  (let ((beg (save-excursion
+               (end-of-line)
+               (re-search-backward elpy-shell-cell-boundary-regexp nil t)
+               (beginning-of-line)
+               (and (string-match-p elpy-shell-codecell-beginning-regexp
+                                    (thing-at-point 'line))
+                    (point))))
+        (end (save-excursion
+               (forward-line)
+               (if (re-search-forward elpy-shell-cell-boundary-regexp nil t)
+                   (forward-line -1)
+                 (end-of-buffer))
+               (end-of-line)
+               (point))))
+    (if beg
+        (progn
+          (elpy-shell--flash-and-message-region beg end)
```

Comment:
Is it not possible to use `elpy-shell-send-region-or-buffer` here to avoid redundancy ?

Reply:
Possible, but then we'd have to store, change, and restore the region. I'd vote to keep it as is. The entire code-sending parts should probabily be simplified into:

1. Navigation commands for start/end
2. Code sending commands (automatically build from navigation commands)

I'll do this at some point. The main reasoning is that users can then also use the navigation commands directly (e.g., to jump to the next codecell etc.)

Reply:
This indeed seems a good idea is the number of functions sending code parts increases.